doc/man/man3/XCProject.3
.TH "XCProject" 3 "Fri Mar 4 2022" "Version 1.1" "XcodeEditor-master" \" -*- nroff -*-
.ad l
.nh
.SH NAME
XCProject
.SH SYNOPSIS
.br
.PP
.PP
Inherits NSObject\&.
.SS "Instance Methods"

.in +1c
.ti -1c
.RI "(id) \- \fBinitWithFilePath:\fP"
.br
.ti -1c
.RI "(id< \fBXcodeGroupMember\fP >) \- \fBgroupMemberWithKey:\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBfiles\fP"
.br
.ti -1c
.RI "(\fBXCSourceFile\fP *) \- \fBfileWithKey:\fP"
.br
.ti -1c
.RI "(\fBXCSourceFile\fP *) \- \fBfileWithName:\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBheaderFiles\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBobjectiveCFiles\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBobjectiveCPlusPlusFiles\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBxibFiles\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBimagePNGFiles\fP"
.br
.ti -1c
.RI "(NSString *) \- \fBfilePath\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCGroup\fP * > *) \- \fBgroups\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBrootGroup\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCGroup\fP * > *) \- \fBrootGroups\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBmainGroup\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBgroupWithKey:\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBgroupWithDisplayName:\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBgroupWithPathFromRoot:\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBgroupForGroupMemberWithKey:\fP"
.br
.ti -1c
.RI "(\fBXCGroup\fP *) \- \fBgroupWithSourceFile:\fP"
.br
.ti -1c
.RI "(void) \- \fBpruneEmptyGroups\fP"
.br
.ti -1c
.RI "(NSArray *) \- \fBversionGroups\fP"
.br
.ti -1c
.RI "(\fBXCVersionGroup\fP *) \- \fBversionGroupWithKey:\fP"
.br
.ti -1c
.RI "(\fBXCVersionGroup\fP *) \- \fBversionGroupWithName:\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCTarget\fP * > *) \- \fBtargets\fP"
.br
.ti -1c
.RI "(\fBXCTarget\fP *) \- \fBtargetWithName:\fP"
.br
.ti -1c
.RI "(NSArray *) \- \fBapplicationTargets\fP"
.br
.ti -1c
.RI "(NSDictionary< NSString *, \fBXCProjectBuildConfig\fP * > *) \- \fBconfigurations\fP"
.br
.ti -1c
.RI "(\fBXCProjectBuildConfig\fP *) \- \fBconfigurationWithName:\fP"
.br
.ti -1c
.RI "(\fBXCProjectBuildConfig\fP *) \- \fBdefaultConfiguration\fP"
.br
.ti -1c
.RI "(void) \- \fBremoveObjectWithKey:\fP"
.br
.ti -1c
.RI "(void) \- \fBsave\fP"
.br
.ti -1c
.RI "(NSMutableDictionary *) \- \fBobjects\fP"
.br
.ti -1c
.RI "(NSMutableDictionary *) \- \fBdataStore\fP"
.br
.ti -1c
.RI "(void) \- \fBdropCache\fP"
.br
.ti -1c
.RI "(NSString *) \- \fBreferenceProxyKeyForName:\fP"
.br
.ti -1c
.RI "(NSArray< \fBXCSourceFile\fP * > *) \- \fBbuildProductsForTargets:\fP"
.br
.ti -1c
.RI "(void) \- \fBaddAsTargetDependency:toTargets:\fP"
.br
.ti -1c
.RI "(NSArray< NSString * > *) \- \fBkeysForProjectObjectsOfType:withIdentifier:singleton:required:\fP"
.br
.ti -1c
.RI "(NSMutableDictionary *) \- \fBPBXProjectDict\fP"
.br
.ti -1c
.RI "(void) \- \fBremoveProxies:\fP"
.br
.ti -1c
.RI "(void) \- \fBaddProxies:\fP"
.br
.ti -1c
.RI "(void) \- \fBremoveFromProjectReferences:forProductsGroup:\fP"
.br
.ti -1c
.RI "(void) \- \fBremoveTargetDependencies:\fP"
.br
.ti -1c
.RI "(NSString *) \- \fBcontainerItemProxyKeyForName:proxyType:\fP"
.br
.ti -1c
.RI "(NSString *) \- \fBproductsGroupKeyForKey:\fP"
.br
.in -1c
.SS "Class Methods"

.in +1c
.ti -1c
.RI "(\fBXCProject\fP *) + \fBprojectWithFilePath:\fP"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "\fBXCFileOperationQueue\fP * \fB_fileOperationQueue\fP"
.br
.ti -1c
.RI "NSString * \fB_filePath\fP"
.br
.ti -1c
.RI "NSMutableDictionary * \fB_dataStore\fP"
.br
.ti -1c
.RI "NSMutableArray * \fB_targets\fP"
.br
.ti -1c
.RI "NSMutableDictionary * \fB_groups\fP"
.br
.ti -1c
.RI "NSMutableDictionary * \fB_versionGroups\fP"
.br
.ti -1c
.RI "NSMutableDictionary * \fB_configurations\fP"
.br
.ti -1c
.RI "NSString * \fB_defaultConfigurationName\fP"
.br
.ti -1c
.RI "NSString * \fB_rootObjectKey\fP"
.br
.in -1c
.SS "Properties"

.in +1c
.ti -1c
.RI "\fBXCFileOperationQueue\fP * \fBfileOperationQueue\fP"
.br
.in -1c
.SH "Method Documentation"
.PP 
.SS "\- (NSArray *) applicationTargets "
Lists the targets which are of application type\&. 
.SS "\- (NSDictionary *) configurations "
Lists the configurations in an xcode project\&. 
.SS "\- (\fBXCProjectBuildConfig\fP *) configurationWithName: (NSString*) name"
Returns the configuration with the specified name, or nil\&. 
.SS "\- (NSArray *) files "
Returns all file resources in the project, as an array of \fC\fBXCSourceFile\fP\fP objects\&. 
.SS "\- (\fBXCSourceFile\fP *) fileWithKey: (NSString*) key"
Returns the project file with the specified key, or nil\&. 
.SS "\- (\fBXCSourceFile\fP *) fileWithName: (NSString*) name"
Returns the project file with the specified name, or nil\&. If more than one project file matches the specified name, which one is returned is undefined\&. 
.SS "\- (\fBXCGroup\fP *) groupForGroupMemberWithKey: (NSString*) key"
Returns the parent _group for the _group or file with the given key; 
.SS "\- (NSArray *) groups "
Lists the groups in an xcode project, returning an array of \fC\fBXCGroup\fP\fP objects\&. 
.SS "\- (\fBXCGroup\fP *) groupWithDisplayName: (NSString*) name"
Returns the \fIfirst\fP group in the project with the given name, or nil\&. 
.SS "\- (\fBXCGroup\fP *) groupWithKey: (NSString*) key"
Returns the group with the given key, or nil\&. 
.SS "\- (\fBXCGroup\fP *) groupWithPathFromRoot: (NSString*) path"
Returns the _group with the specified display name path - the directory relative to the root _group\&. Eg Source/Main 
.SS "\- (\fBXCGroup\fP *) groupWithSourceFile: (\fBXCSourceFile\fP*) sourceFile"
Returns the parent group for the group or file with the source file 
.SS "\- (NSArray *) headerFiles "
Returns all header files in the project, as an array of \fC\fBXCSourceFile\fP\fP objects\&. 
.SS "\- (id) initWithFilePath: (NSString*) filePath"
Creates a new project editor instance with the specified Project\&.xcodeproj file\&. 
.SS "\- (\fBXCGroup\fP *) mainGroup "
Returns the main group under root object 
.SS "\- (NSArray *) objectiveCFiles "
Returns all implementation obj-c implementation files in the project, as an array of \fC\fBXCSourceFile\fP\fP objects\&. 
.SS "\- (NSArray *) objectiveCPlusPlusFiles "
Returns all implementation obj-c++ implementation files in the project, as an array of \fC\fBXCSourceFile\fP\fP objects\&. 
.SS "\- (NSMutableDictionary *) objects "
Raw project data\&. 
.SS "\- (void) pruneEmptyGroups "
Removes all empty groups from the project\&. 
.SS "\- (\fBXCGroup\fP *) rootGroup "
Returns the root (top-level) _group\&. 
.SS "\- (NSArray *) rootGroups "
Returns the root (top-level) groups, if there are multiple\&. An array of rootGroup if there is only one\&. 
.SS "\- (void) save "
Saves a project after editing\&. 
.SS "\- (NSArray *) targets "
Lists the targets in an xcode project, returning an array of \fC\fBXCTarget\fP\fP objects\&. 
.SS "\- (\fBXCTarget\fP *) targetWithName: (NSString*) name"
Returns the target with the specified name, or nil\&. 
.SS "\- (NSArray *) versionGroups "
Lists the version groups in an xcode project, returning an array of \fC\fBXCVersionGroup\fP\fP objects\&. 
.SS "\- (\fBXCVersionGroup\fP *) versionGroupWithKey: (NSString*) key"
Returns the version group with the given key, or nil\&. 
.SS "\- (\fBXCVersionGroup\fP *) versionGroupWithName: (NSString*) name"
Returns the version group with the given file name, or nil\&. 
.SS "\- (NSArray *) xibFiles "
Returns all the xib files in the project, as an array of \fC\fBXCSourceFile\fP\fP objects\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for XcodeEditor-master from the source code\&.


Source/XCProject.m
////////////////////////////////////////////////////////////////////////////////
//
//  JASPER BLUES
//  Copyright 2012 - 2013 Jasper Blues
//  All Rights Reserved.
//
//  NOTICE: Jasper Blues permits you to use, modify, and distribute this file
//  in accordance with the terms of the license agreement accompanying it.
//
////////////////////////////////////////////////////////////////////////////////


#import "XCProject.h"
#import "XCGroup.h"
#import "XCSourceFile.h"
#import "XCTarget.h"
#import "XCFileOperationQueue.h"
#import "XCProjectBuildConfig.h"
#import "XCVersionGroup.h"
#import "XcodeEditor-Prefix.pch"

NSString *const XCProjectNotFoundException;

@implementation XCProject


@synthesize fileOperationQueue = _fileOperationQueue;

//-------------------------------------------------------------------------------------------
#pragma mark - Class Methods
//-------------------------------------------------------------------------------------------

+ (XCProject *)projectWithFilePath:(NSString *)filePath
{
    return [[XCProject alloc] initWithFilePath:filePath];
}


//-------------------------------------------------------------------------------------------
#pragma mark - Initialization & Destruction
//-------------------------------------------------------------------------------------------

- (id)initWithFilePath:(NSString *)filePath
{
    if ((self = [super init])) {
        _filePath = [filePath copy];
        _dataStore = [[NSMutableDictionary alloc]
                initWithContentsOfFile:[_filePath stringByAppendingPathComponent:@"project.pbxproj"]];

        if (!_dataStore) {
            [NSException raise:XCProjectNotFoundException format:@"Project file not found at file path %@", _filePath];
        }

        _fileOperationQueue =
                [[XCFileOperationQueue alloc] initWithBaseDirectory:[_filePath stringByDeletingLastPathComponent]];

    }
    return self;
}

//-------------------------------------------------------------------------------------------
#pragma mark - Interface Methods
//-------------------------------------------------------------------------------------------

#pragma mark General Group member

- (id<XcodeGroupMember>)groupMemberWithKey:(NSString *)key
{
    NSDictionary *obj = [[self objects] valueForKey:key];
    
    if (obj) {
        NSString *groupIsa =[obj valueForKey:@"isa"];
        if([groupIsa xce_hasFileReferenceOrReferenceProxyType]) {
            return [self fileWithKey:key];
        }
        else if([groupIsa xce_hasVersionedGroupType]) {
            return [self versionGroupWithKey:key];
        }
        else if([groupIsa xce_hasGroupType]) {
            return [self groupWithKey:key];
        }
    }
    return nil;
}

#pragma mark Files

- (NSArray *)files
{
    NSMutableArray *results = [NSMutableArray array];
    [[self objects] enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSDictionary *obj, BOOL *stop) {
        if ([[obj valueForKey:@"isa"] xce_hasFileReferenceType]) {
            XcodeSourceFileType fileType = XCSourceFileTypeFromStringRepresentation(
                    [obj valueForKey:@"lastKnownFileType"] ?: [obj valueForKey:@"explicitFileType"]);
            NSString *path = [obj valueForKey:@"path"];
            NSString *sourceTree = [obj valueForKey:@"sourceTree"];
            XCSourceFile *sourceFile = [XCSourceFile sourceFileWithProject:self key:key type:fileType name:path
                                                                sourceTree:(sourceTree ?: @"<_group>") path:nil];
            [results addObject:sourceFile];
        }
    }];
    return results;
}

- (XCSourceFile *)fileWithKey:(NSString *)key
{
    NSDictionary *obj = [[self objects] valueForKey:key];
    if (obj && [[obj valueForKey:@"isa"] xce_hasFileReferenceOrReferenceProxyType]) {
        XcodeSourceFileType fileType = XCSourceFileTypeFromStringRepresentation(
                [obj valueForKey:@"lastKnownFileType"] ?: [obj valueForKey:@"explicitFileType"]);

        NSString *name = [obj valueForKey:@"name"];
        NSString *sourceTree = [obj valueForKey:@"sourceTree"];
        NSString *path = [obj valueForKey:@"path"];

        if (name == nil) {
            name = path;
        }
        return [XCSourceFile sourceFileWithProject:self key:key type:fileType name:name
                                        sourceTree:(sourceTree ?: @"<_group>") path:path];
    }
    return nil;
}

- (XCSourceFile *)fileWithName:(NSString *)name
{
    for (XCSourceFile *projectFile in [self files]) {
        if ([[projectFile name] isEqualToString:name]) {
            return projectFile;
        }
    }
    return nil;
}

- (NSArray *)headerFiles
{
    return [self projectFilesOfType:SourceCodeHeader];
}

- (NSArray *)objectiveCFiles
{
    return [self projectFilesOfType:SourceCodeObjC];
}

- (NSArray *)objectiveCPlusPlusFiles
{
    return [self projectFilesOfType:SourceCodeObjCPlusPlus];
}


- (NSArray *)xibFiles
{
    return [self projectFilesOfType:XibFile];
}

- (NSArray *)imagePNGFiles
{
    return [self projectFilesOfType:ImageResourcePNG];
}


// need this value to construct relative path in XcodeprojDefinition
- (NSString *)filePath
{
    return _filePath;
}

//-------------------------------------------------------------------------------------------
#pragma mark Groups
//-------------------------------------------------------------------------------------------

- (NSArray *)groups
{
    NSMutableArray *results = [[NSMutableArray alloc] init];
    [[self objects] enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSDictionary *obj, BOOL *stop) {
        if ([[obj valueForKey:@"isa"] xce_hasGroupType]) {
            XCGroup *group = _groups[key];
            if (group == nil) {
                group = [self createGroupWithDictionary:obj forKey:key];
                _groups[key] = group;
            }
            [results addObject:group];
        }
    }];
    return results;
}

//TODO: Optimize this implementation.
- (XCGroup *)rootGroup
{
    for (XCGroup *group in [self groups]) {
        if ([group isRootGroup]) {
            return group;
        }
    }
    return nil;
}

- (NSArray *)rootGroups
{
    XCGroup *group = [self rootGroup];
    if (group) {
        return [NSArray arrayWithObject:group];
    }

    NSMutableArray *results = [NSMutableArray array];
    for (XCGroup *group in [self groups]) {
        if ([group parentGroup] == nil) {
            [results addObject:group];
        }
    }

    return [results copy];
}

- (XCGroup *)mainGroup
{
    NSString* rootObjectKey = [self rootObjectKey];
    NSDictionary* rootObject = [[self objects] objectForKey:rootObjectKey];
    NSString* mainGroupKey = [rootObject objectForKey:@"mainGroup"];
    for (XCGroup* group in [self groups]) {
        if ([group.key isEqualToString:mainGroupKey])
             return group;
    }
             
    return nil;
}

- (XCGroup *)groupWithKey:(NSString *)key
{
    XCGroup *group = [_groups objectForKey:key];
    if (group) {
        return group;
    }

    NSDictionary *obj = [[self objects] objectForKey:key];
    if (obj && [[obj valueForKey:@"isa"] xce_hasGroupType]) {
        XCGroup *group = [self createGroupWithDictionary:obj forKey:key];
        _groups[key] = group;

        return group;
    }
    return nil;
}

- (XCGroup *)groupWithDisplayName:(NSString *)name
{
    for (XCGroup *group in [self groups]) {
        if ([[group displayName] isEqualToString:name]) {
            return group;
        }
    }
    return nil;
}


- (XCGroup *)groupForGroupMemberWithKey:(NSString *)key
{
    for (XCGroup *group in [self groups]) {
        if ([group memberWithKey:key]) {
            return group;
        }
    }
    return nil;
}

- (XCGroup *)groupWithSourceFile:(XCSourceFile *)sourceFile
{
    for (XCGroup *group in [self groups]) {
        for (id <XcodeGroupMember> member in [group members]) {
            if ([member isKindOfClass:[XCSourceFile class]] && [[sourceFile key] isEqualToString:[member key]]) {
                return group;
            }
        }
    }
    return nil;
}

- (void)pruneEmptyGroups
{
    [self doPruneEmptyGroups];
}


//TODO: This could fail if the path attribute on a given group is more than one directory. Start with candidates and
//TODO: search backwards.
- (XCGroup *)groupWithPathFromRoot:(NSString *)path
{
    NSArray *pathItems = [path pathComponents];
    XCGroup *currentGroup = [self rootGroup];
    for (NSString *pathItem in pathItems) {
        id <XcodeGroupMember> group = [currentGroup memberWithDisplayName:pathItem];
        if ([group isKindOfClass:[XCGroup class]]) {
            currentGroup = (XCGroup*)group;
        } else {
            return nil;
        }
    }
    return currentGroup;
}

- (XCGroup *)createGroupWithDictionary:(NSDictionary *)dictionary forKey:(NSString *)key
{
    return [XCGroup groupWithProject:self
                                 key:key
                               alias:[dictionary valueForKey:@"name"]
                                path:[dictionary valueForKey:@"path"]
                            children:[dictionary valueForKey:@"children"]
            memberType:[[dictionary valueForKey:@"isa"] xce_asMemberType]];
}

//-------------------------------------------------------------------------------------------
#pragma mark version group
//-------------------------------------------------------------------------------------------

- (NSArray*)versionGroups
{
    NSMutableArray* results = [[NSMutableArray alloc] init];
    [[self objects] enumerateKeysAndObjectsUsingBlock:^(NSString* key, NSDictionary* obj, BOOL* stop)
     {
         if ([[obj valueForKey:@"isa"] xce_hasVersionedGroupType])
         {
             XCVersionGroup* group = _versionGroups[key];
             if (group == nil)
             {
                 group = [self createVersionGroupWithDictionary:obj forKey:key];
                 _versionGroups[key] = group;
             }
             [results addObject:group];
         }
     }];
    return results;
}

- (XCVersionGroup*)versionGroupWithKey:(NSString*)key
{
    XCVersionGroup* group = [_versionGroups objectForKey:key];
    if (group)
    {
        return group;
    }
    
    NSDictionary* obj = [[self objects] objectForKey:key];
    if (obj && [[obj valueForKey:@"isa"] xce_hasVersionedGroupType])
    {
        XCVersionGroup* group = [self createVersionGroupWithDictionary:obj forKey:key];
        _versionGroups[key] = group;
        
        return group;
    }
    return nil;
}

- (XCVersionGroup *)versionGroupWithName:(NSString *)name
{
    for (XCVersionGroup* group in [self versionGroups])
    {
        if([[[group pathRelativeToParent] stringByDeletingPathExtension]isEqualToString:name])
            return group;
    }
    return nil;
}

- (XCVersionGroup*)createVersionGroupWithDictionary:(NSDictionary*)dictionary forKey:(NSString*)key
{
    return [XCVersionGroup versionGroupWithProject:self
                                               key:key
                                              path:[dictionary valueForKey:@"path"]
                                          children:[dictionary valueForKey:@"children"]
                                    currentVersion:[dictionary valueForKey:@"currentVersion"]];
}

//-------------------------------------------------------------------------------------------
#pragma mark targets
//-------------------------------------------------------------------------------------------

- (NSArray *)targets
{
    if (_targets == nil) {
        _targets = [[NSMutableArray alloc] init];
        [[self objects] enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSDictionary *obj, BOOL *stop) {
            if ([[obj valueForKey:@"isa"] xce_hasNativeTargetType]) {
                XCTarget *target = [XCTarget targetWithProject:self key:key name:[obj valueForKey:@"name"]
                                                   productName:[obj valueForKey:@"productName"]
                                              productReference:[obj valueForKey:@"productReference"]
                                              productType:[obj valueForKey:@"productType"]];
                [_targets addObject:target];
            }
        }];
    }
    return _targets;
}

- (XCTarget *)targetWithName:(NSString *)name
{
    for (XCTarget *target in [self targets]) {
        if ([[target name] isEqualToString:name]) {
            return target;
        }
    }
    return nil;
}

- (NSArray*)applicationTargets
{
    NSArray* targets = [self targets];
    NSArray* filteredTargets = [targets filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id  _Nonnull evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
        return [evaluatedObject isApplicationType];
    }]];
    
    return filteredTargets;
}

- (NSData *)_fixEncodingInData:(NSData *)data
{
    NSString *source = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    NSMutableString *destination = @"".mutableCopy;
    for(int i = 0; i < source.length; i++) {
        unichar c = [source characterAtIndex:i];
        if(c < 128) {
            [destination appendFormat:@"%c", c];
        } else {
            [destination appendFormat:@"&#%u;", c];
        }
    }
    
    return [destination dataUsingEncoding:NSUTF8StringEncoding];
}

- (void)save
{
    [_fileOperationQueue commitFileOperations];

    NSData *data = [NSPropertyListSerialization dataWithPropertyList:_dataStore format:NSPropertyListXMLFormat_v1_0 options:0 error:NULL];
    data = [self _fixEncodingInData:data];
    [data writeToFile:[_filePath stringByAppendingPathComponent:@"project.pbxproj"] atomically:YES];
    
    // Don't forget to reset the cache so that we'll always get the latest data.
    [self dropCache];
    if ( DEBUG ) printf("Project %s saved\n", _filePath.lastPathComponent.UTF8String);
}

- (NSMutableDictionary *)objects
{
    return [_dataStore objectForKey:@"objects"];
}

- (NSMutableDictionary *)dataStore
{
    return _dataStore;
}

- (void)dropCache
{
    _targets = nil;
    _configurations = nil;
    _rootObjectKey = nil;
}


- (NSDictionary *)configurations
{
    if (_configurations == nil) {
        NSString *buildConfigurationRootSectionKey =
                [[[self objects] objectForKey:[self rootObjectKey]] objectForKey:@"buildConfigurationList"];
        NSDictionary *buildConfigurationDictionary = [[self objects] objectForKey:buildConfigurationRootSectionKey];
        _configurations =
                [[XCProjectBuildConfig buildConfigurationsFromArray:[buildConfigurationDictionary objectForKey:@"buildConfigurations"]
                                                          inProject:self] mutableCopy];
        _defaultConfigurationName = [[buildConfigurationDictionary objectForKey:@"defaultConfigurationName"] copy];
    }

    return [_configurations copy];
}

- (XCProjectBuildConfig *)configurationWithName:(NSString *)name
{
    return [[self configurations] objectForKey:name];
}

- (XCProjectBuildConfig *)defaultConfiguration
{
    return [[self configurations] objectForKey:_defaultConfigurationName];
}

//-------------------------------------------------------------------------------------------
#pragma mark Deletion

- (void)removeObjectWithKey:(NSString*)key
{
    if([self.objects valueForKey:key])
    {
        XCGroup *group;
        if((group = [self groupForGroupMemberWithKey:key])!=nil)
        {
            [group removeMemberWithKey:key];
        }
        [self.objects removeObjectForKey:key];
    }
}

//-------------------------------------------------------------------------------------------
#pragma mark Private
//-------------------------------------------------------------------------------------------

- (NSString *)rootObjectKey
{
    if (_rootObjectKey == nil) {
        _rootObjectKey = [[_dataStore objectForKey:@"rootObject"] copy];;
    }

    return _rootObjectKey;
}

- (NSArray *)projectFilesOfType:(XcodeSourceFileType)projectFileType
{
    NSMutableArray *results = [NSMutableArray array];
    for (XCSourceFile *file in [self files]) {
        if ([file type] == projectFileType) {
            [results addObject:file];
        }
    }
    return results;
}

- (BOOL)doPruneEmptyGroups
{
    BOOL hadEmptyGroups = NO;
    for (XCGroup *group in [self groups]) {
        if ([group isEmpty]) {
            hadEmptyGroups = YES;
            [group removeFromParentGroup];
        }
    }
    //Prune any groups made empty as a result of pruning a child group.
    if (hadEmptyGroups) {
        [self doPruneEmptyGroups];
    }
    return hadEmptyGroups;
}

@end
